navigation_and_routing/lib/src/app.dart

```diff
@@ -16,98 +28,245 @@ class Bookstore extends StatefulWidget {
-      routeState: _routeState,
-      navigatorKey: _navigatorKey,
-      builder: (context) => BookstoreNavigator(
-        navigatorKey: _navigatorKey,
-      ),
-    );
-
-    // Listen for when the user logs out and display the signin screen.
-    _auth.addListener(_handleAuthStateChanged);
-
-    super.initState();
-  }
+  final BookstoreAuth auth = BookstoreAuth();
 
   @override
-  Widget build(BuildContext context) => RouteStateScope(
-        notifier: _routeState,
-        child: BookstoreAuthScope(
-          notifier: _auth,
-          child: MaterialApp.router(
-            routerDelegate: _routerDelegate,
-            routeInformationParser: _routeParser,
-            // Revert back to pre-Flutter-2.5 transition behavior:
-            // https://github.com/flutter/flutter/issues/82053
-            theme: ThemeData(
-              useMaterial3: true,
-              pageTransitionsTheme: const PageTransitionsTheme(
-                builders: {
-                  TargetPlatform.android: FadeUpwardsPageTransitionsBuilder(),
-                  TargetPlatform.iOS: CupertinoPageTransitionsBuilder(),
-                  TargetPlatform.linux: FadeUpwardsPageTransitionsBuilder(),
-                  TargetPlatform.macOS: CupertinoPageTransitionsBuilder(),
-                  TargetPlatform.windows: FadeUpwardsPageTransitionsBuilder(),
+  Widget build(BuildContext context) {
+    return MaterialApp.router(
+      builder: (context, child) {
+        if (child == null) {
+          throw ('No child in .router constructor builder');
+        }
+        return BookstoreAuthScope(
+          notifier: auth,
+          child: child,
+        );
+      },
+      routerConfig: GoRouter(
+        refreshListenable: auth,
+        debugLogDiagnostics: true,
+        initialLocation: '/books/popular',
+        redirect: (context, state) {
+          final signedIn = BookstoreAuth.of(context).signedIn;
+          if (state.uri.toString() != '/sign-in' && !signedIn) {
+            return '/sign-in';
+          }
+          return null;
+        },
+        routes: [
+          ShellRoute(
+            navigatorKey: appShellNavigatorKey,
+            builder: (context, state, child) {
+              return BookstoreScaffold(
+                child: child,
+                selectedIndex: switch (state.uri.path) {
+                  var p when p.startsWith('/books') => 0,
+                  var p when p.startsWith('/authors') => 1,
+                  var p when p.startsWith('/settings') => 2,
+                  _ => 0,
+                },
+              );
+            },
+            routes: [
+              ShellRoute(
+                pageBuilder: (context, state, child) {
+                  return FadeTransitionPage<dynamic>(
+                    key: state.pageKey,
+                    // Use a builder to get the correct BuildContext
+                    // TODO (johnpryan): remove when https://github.com/flutter/flutter/issues/108177 lands
```

Comment:
Consider raising an issue in `flutter/samples` issue tracker to keep this task visible.

Reply:
https://github.com/flutter/samples/issues/2074